Erm 230 editorgs

aditya-matukumalli opened the pull request:
- Added new _OrganizationsFieldArray_ component that renders an __OrganizationCard__ component along with organization role selector.
- Added new _OrganizationCard_ component used to render interfaces/orgName/default card layouts conditionally.
- Added new _onMarkforDeletion_ prop to WithKiwtFieldArray that marks a field for deletion.

doytch commented on lib/OrganizationCard/OrganizationCard.js:
```diff
@@ -0,0 +1,115 @@
+      });
+
+      return (
+        <React.Fragment>
```
Why return this in a Fragment?

doytch commented on lib/OrganizationCard/OrganizationCard.js:
```diff
@@ -0,0 +1,115 @@
+            />}
+          >
+            <React.Fragment>
+              <InterfacesCard name={name} index={index} organization={organization} {...this.props} />
```
Also, don't pass down `this.props` like this carelessly. Why would an InterfacesCard need the `onAddField` or `onDeleteField` props?

doytch commented on package.json:
```diff
@@ -43,6 +43,7 @@
     "prop-types": "^15.6.0",
     "react-dropzone": "^10.0.0",
     "react-intl": "^2.4.0",
+    "react-router-dom": "^5.0.1",
```
Why?

doytch commented on translations/stripes-erm-components/en.json:
```diff
@@ -35,7 +35,10 @@
+  "organizations.noOrganization": "No organizations added",
+  "organizations.addOrganizationToStart": "Add an organization to get started",
+  "organizations.addOrganizationToAgreement": "Add organization to agreement",
+  "organizations.unlinkOrganization": "Unlink organization",
```
Please be mindful of whitespace. There's a reason there's an empty line between each section.

doytch commented on lib/OrganizationCard/InterfacesCard.js:
```diff
@@ -0,0 +1,182 @@
+        headerEnd={this.renderOrganizationPicker(index, name)}
+        roundedBorder
+      >
+        <React.Fragment>
```
Why is this a Fragment? Don't overuse Fragments please.

doytch commented on lib/OrganizationCard/InterfacesCard.js:
```diff
@@ -0,0 +1,182 @@
+} from '@folio/stripes/components';
+import OrganizationPicker from './OrganizationPicker';
+
+export default class InterfacesCard extends React.Component {
```
So this isn't really just an InterfacesCard is it? It renders the entire Organization card...and may list no organizations whatsoever. Can we clean up the way all these components relate to one another? It's super confusing.

doytch replied:
Also, this component is too big. It renders two different kinds of cards that do different things because one of them is in a Field and one isn't.

id-jenkins commented:
<html><body><pre>
yarn run v1.9.4
$ eslint .

/home/jenkins/workspace/org_stripes-erm-components_PR-77/project/lib/OrganizationCard/InterfacesCard.js
  26:13  error  'organizationPickerComponent' is missing in props validation  react/prop-types

✖ 1 problem (1 error, 0 warnings)

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

id-jenkins commented:
<html><body><pre>
yarn run v1.9.4
$ eslint .

/home/jenkins/workspace/org_stripes-erm-components_PR-77/project/lib/OrganizationsFieldArray/OrgInfo.js
  20:34  error  'organizationPickerComponent' PropType is defined but prop is never used  react/no-unused-prop-types

✖ 1 problem (1 error, 0 warnings)

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

doytch commented on lib/OrganizationsFieldArray/OrganizationsFieldArray.js:
```diff
@@ -1,144 +1,72 @@
+        return (
+          <OrganizationCard
+            organization={organization}
+            index={index}
```
TODO: Issue?

doytch commented on lib/OrganizationsFieldArray/OrganizationsFieldArray.js:
```diff
@@ -1,144 +1,72 @@
+          <OrganizationCard
+            organization={organization}
+            index={index}
+            {...this.props}
```
Is this necessary? Pass down only what's needed.

doytch commented on lib/OrganizationsFieldArray/OrganizationsFieldArray.js:
```diff
@@ -1,144 +1,72 @@
-  render() {
-    const { items, onAddField } = this.props;
+      return (
+        <React.Fragment>
```
Why is this in a Fragment?

doytch commented on lib/OrganizationsFieldArray/OrganizationCard.js:
```diff
@@ -0,0 +1,60 @@
+        data-test-organizations-org
+        hasMargin
+        key={index}
+        headerStart={<strong>{`Organization ${index + 1}`}</strong>}
```
i18n

doytch commented on lib/OrganizationsFieldArray/OrganizationCard.js:
```diff
@@ -0,0 +1,60 @@
+} from '@folio/stripes/components';
+import OrgInfo from './OrgInfo';
+
+export default class OrganizationCard extends React.Component {
```
It might make sense to call this something like `EditOrganizationCard`, since we're gonna have a card within it that should be the one that's _actually_ called `OrganizationCard`.

doytch commented on lib/OrganizationsFieldArray/OrganizationCard.js:
```diff
@@ -0,0 +1,60 @@
+    onDeleteField: PropTypes.func,
+    onReplaceField: PropTypes.func,
+    onAddField: PropTypes.func,
+    organizationPickerComponent: PropTypes.elementType,
```
Is it possible to remove this whole `organizationPickerComponent` thing? Would it simplify this whole set of components if we removed it since it looks like we're passing it down several levels? It was only there to get higher coverage stats and it didn't really work out.

doytch commented on lib/OrganizationsFieldArray/OrgInfo.js:
```diff
@@ -0,0 +1,162 @@
+} from '@folio/stripes/components';
+import { required } from '../validators';
+
+export default class OrgInfo extends React.Component {
```
Could we change this component to:
1. Not do the decision making of whether to render the org info vs org picker, moving that logic up to the `EditOrganizationCard`. So of its current responsibilities, remove everything that isn't rendering the org name and role.
1. Support rendering of interfaces if they are passed into it as a prop.
1. Rename this to `OrganizationCard`

?

If so, could we then use this component in _both_ the `EditOrganizationCard` as well as where we're currently rendering out the organizations in `ui-agreements/src/components/AgreementSections/Organizations.js`? That would be nice.

doytch replied:
In any case, the decision making that I say in bullet 1 should probably be moved up to that hypothetical `EditOrganizationCard` (what's currently just called `OrganizationCard`). This design results in a lot of props passing when it wouldn't be strictly necessary.

doytch commented on lib/OrganizationsFieldArray/OrgInfo.js:
```diff
@@ -0,0 +1,162 @@
+  renderDefault = () => {
+    const { index, name } = this.props;
+    return (
+      <Field
```
Why is this a Field rendering a Card? Card has no handling for any of the props that redux-form will give it.

doytch commented on lib/OrganizationsFieldArray/OrgInfo.js:
```diff
@@ -0,0 +1,162 @@
+    this.props.onReplaceField(index, { org });
+  }
+
+  unlinkOrg = (index, orgId) => {
```
This definitely needs comments explaining what's happening here.

doytch commented on lib/OrganizationsFieldArray/OrgInfo.js:
```diff
@@ -0,0 +1,162 @@
+      <span>
+        <AppIcon
+          app="organizations"
+          appIconKey="organizations"
```
Is `appIconKey` necessary here?

doytch commented on lib/OrganizationsFieldArray/OrgInfo.js:
```diff
@@ -0,0 +1,162 @@
+        cardStyle="positive"
+        hasMargin
+        headerStart={
+          <strong>Organization</strong>
```
i18n. Can you do a pass through this entire PR checking for untranslated strings?

doytch commented on lib/OrganizationsFieldArray/OrgInfo.js:
```diff
@@ -0,0 +1,162 @@
+          <span>
+            <AppIcon
+              app="organizations"
+              appIconKey="organizations"
```
Is `appIconKey` necessary here?

doytch commented on lib/OrganizationsFieldArray/OrganizationsFieldArray.js:
```diff
@@ -1,144 +1,83 @@
-  render() {
-    const { items, onAddField } = this.props;
+    renderOrganizations = (items) => {
+      const { onMarkforDeletion, name, roles, organizationPickerComponent, onDeleteField, onReplaceField } = this.props;
```
Also, we've been trying to always alphabetize destructuring statements and prop lists. There's a few examples of that throughout this PR. It just makes it a lot easier to skim the code later.

doytch commented on lib/OrganizationsFieldArray/OrganizationsFieldArray.js:
```diff
@@ -1,144 +1,83 @@
+            name={name}
+            organization={organization}
+            index={index}
+            items={items}
```
Can you please be careful and conscious about the prop passing you're doing? I don't see why `items` needs to be passed in, and it's not declared in the prop types for `EditOrganizationCard`. This kind of stuff has come up a lot of times during this PR and it's tough to constantly scroll back and forth through Github trying to figure whether I'm missing something or not.

doytch commented on lib/OrganizationsFieldArray/EditOrganizationCard.js:
```diff
@@ -0,0 +1,229 @@
+    <Field
+      data-test-org-name
+      component={OrganizationCard}
+      {...this.getOrganizationCardProps()}
```
Is it really necessary to define this in a separate function? I don't see it being used anywhere else...

doytch commented on lib/OrganizationsFieldArray/EditOrganizationCard.js:
```diff
@@ -0,0 +1,229 @@
+
+  renderHeaderStart = (isOrgPresent) => {
+    return (
+      isOrgPresent ?
```
Could we just always render the same thing here, with the `AppIcon`?

doytch commented on lib/OrganizationsFieldArray/EditOrganizationCard.js:
```diff
@@ -0,0 +1,229 @@
+  renderHeaderEnd = (isOrgPresent, id) => {
+    const { index, name } = this.props;
+    return isOrgPresent ?
+      this.renderOrgInfoHeaderEnd(id) :
```
Seems like these two functions don't really need to be defined in separate functions. It just leads to inconsistencies like passing `id` to `renderOrgInfoHeaderEnd` when it doesn't use that param.

aditya-matukumalli replied:
Done

doytch commented on lib/OrganizationsFieldArray/EditOrganizationCard.js:
```diff
@@ -0,0 +1,229 @@
+    </div>
+  );
+
+  onAddOrg = (organization, index) => {
```
We've been trying to keep a pattern of callback handlers being called `handleFoo` and being passed as `onFoo`. So this would be called `handleAddOrg` and the below would be called `handleUnlinkOrg`. It makes it easier to track where something is defined rather than where it's just being used or passed through.

doytch commented on lib/OrganizationsFieldArray/OrganizationCard.js:
```diff
@@ -0,0 +1,134 @@
+    renderDefault: PropTypes.func,
+  };
+
+  renderError = (error) => (
```
This function isn't called anywhere.

doytch commented on lib/OrganizationsFieldArray/OrganizationCard.js:
```diff
@@ -0,0 +1,134 @@
+    ]),
+    id: PropTypes.string,
+    interfaces: PropTypes.object,
+    meta: PropTypes.shape({
```
With this and the way you're doing the error handling, you're expressing the dependency that `Card` should a redux form field. This isn't ideal and we don't even always use it like that ourselves, like when we're on the view agreement/license pane.

doytch replied:
I guess this also loops back to the above comment talking about whether "this component [should] be used when rendering both a selected and unselected organization". 

If it was only used when rendering a _selected_ organization, then you wouldn't need handle displaying errors in this component since the "role" dropdown could always be responsible for it. Then you wouldn't need this dependency on `meta` supplied by redux-form.

doytch commented on lib/OrganizationsFieldArray/OrganizationCard.js:
```diff
@@ -0,0 +1,134 @@
+          notes: item => get(item, 'notes'),
+          username: item => get(item, 'username'),
+          password: item => get(item, 'password'),
+          type: item => get(item, 'type'),
```
[`type` is an array](https://s3.amazonaws.com/foliodocs/api/mod-organizations-storage/interface.html#organizations_storage_interfaces__id__get)

doytch commented on lib/OrganizationsFieldArray/OrganizationCard.js:
```diff
@@ -0,0 +1,134 @@
+  render() {
+    const {
+      org,
+      renderDefault,
```
Can we call this `renderEmpty`? "Default" is pretty ambiguous. Ie, default compared to what?

doytch commented on lib/OrganizationsFieldArray/EditOrganizationCard.js:
```diff
@@ -0,0 +1,229 @@
+    return (
+      <Button
+        data-test-unlink-button
+        buttonStyle="danger"
```
This button is misaligned vertically because of bottom margin. [Apply `marginBottom0` to it.](https://github.com/folio-org/stripes-components/tree/master/lib/Button#props)

Same thing with the "Add Organization" button when an org hasn't been selected.

aditya-matukumalli replied:
_Add Organization_ Button is rendered here https://github.com/folio-org/ui-plugin-find-organization/blob/master/OrganizationSearch/OrganizationSearch.js#L51 and looks like it doens't have a capability to render a _marginBottom0_ style prop. Might have to raise a PR against ui-plugin0find-org and add it.

aditya-matukumalli replied:
Raised a PR for this against the plugin

doytch commented on lib/OrganizationsFieldArray/OrganizationCard.js:
```diff
@@ -0,0 +1,134 @@
+
+export default class OrganizationCard extends React.Component {
+  static propTypes = {
+    cardStyle: PropTypes.string,
```
I get why you're passing `cardStyle` in here but the whole point of this card is so that we don't have to remember what particular props to set on `Card` to have the right styling. At the very least, this should probably have a default prop value so that common usage of rendering an organization doesn't require `cardStyle` to be passed in.

It's part of what makes me unsure whether it's actually appropriate for this component to be used when rendering both a selected and unselected organization. I kind of feel like this should only be used for a _selected_ organization. We can talk about this if you want.

aditya-matukumalli commented:
@doytch 
- Added the documentation readme files
- Moved _ViewOrganizationCard_ into its own file
- Added Tests.

I wasn't able to make use of the Plugin in the tests somehow. So just used the TextField as we did previously in the tests as is in there for the _OrganiztionsFieldArray._ So, It Didn't make sense to write tests for _EditOrganizationCard_ separately. Added in few tests for the _ViewOrganizationCard_ component.

doytch commented on lib/ViewOrganizationCard/tests/ViewOrganizationCard-test.js:
```diff
@@ -0,0 +1,72 @@
+    expect(interactor.hasCardBody).to.be.true;
+  });
+
+  it('renders a Multi column list', () => {
```
Can we check the validity of this list? Eg, some checks for the contents matching the interfaces we're passing in?

aditya-matukumalli replied:
Yup, will add that in

doytch commented on lib/ViewOrganizationCard/ViewOrganizationCard.js:
```diff
@@ -0,0 +1,110 @@
+        />
+        &nbsp;
+        <strong id={id} data-test-org-name>
+          {`${orgName}`}
```
This can just be `{orgName}`

doytch commented on lib/OrganizationsFieldArray/SelectOrganizationCard.js:
```diff
@@ -0,0 +1,57 @@
+          headerEnd={headerEnd}
+          roundedBorder
+        >
+          { error ? this.renderError(error) : (renderEmpty && renderEmpty())}
```
It's up to you, but would it make more sense to just use `this.props.children` instead of `renderEmpty` for this kind of thing? You could always render `this.props.children` to make sure any instructions in there are visible even when there's an error as well?

doytch commented on lib/OrganizationsFieldArray/SelectOrganizationCard.js:
```diff
@@ -0,0 +1,56 @@
+
+export default class SelectOrganizationCard extends React.Component {
+  static propTypes = {
+    children: PropTypes.func,
```
```suggestion
    children: PropTypes.node,
```
